### PR TITLE
fix `ConfusionMatrix` and `StatScores` for num_classes > 16

### LINKED ISSUE #1510: MulticlassConfusionMatrix overflows for num_classes > 16
## 🐛 Bug

_multiclass_confusion_matrix_update() overflows for num_classes > 16, target.dtype=uint8 and preds.max()>15

### To Reproduce

```
import torch
from torchmetrics.classification import MulticlassConfusionMatrix

m = MulticlassConfusionMatrix(num_classes=17)
preds = torch.LongTensor([16])
targets = torch.ByteTensor([16])

m.update(preds, targets)
t = m.compute()
assert t[16, 16] == 1
assert t[1, 15] == 0
```

### Expected behavior

t[16, 16] == 1
t[1, 15] == 0

### Environment

- TorchMetrics version : fork master

### Additional context

Probably the same issue for **MulticlassStatScore**
```
def _multiclass_confusion_matrix_update(preds: Tensor, target: Tensor, num_classes: int) -> Tensor:
    """Compute the bins to update the confusion matrix with."""
    -- unique_mapping = (target * num_classes + preds).to(torch.long)
    ++ unique_mapping = target.to(torch.long) * num_classes + preds.to(torch.long)
    bins = _bincount(unique_mapping, minlength=num_classes**2)
    return bins.reshape(num_classes, num_classes)
```

### PR DESCRIPTION
**Pull request by vincentvaroquauxads**
e.g. if preds or target is uint8 and num_classes > 16, unique_mapping overflows

## What does this PR do?

Fixes #1510

I would suggest to define :
``` unittests.helpers.NUM_CLASSES=17```
and add something like the following to `unittests.classification.inputs._multiclass_cases`
```py
    pytest.param(
        Input(
            preds=_logsoftmax(torch.rand(NUM_BATCHES, BATCH_SIZE, NUM_CLASSES), -1),
            target=torch.randint(high=NUM_CLASSES, size=(NUM_BATCHES, BATCH_SIZE), dtype=torch.uint8),
        ),
        id="input[single dim uint8-logits]",
    ),
```
But it breaks a lot of tests...

### CODE REVIEW

**Comment by justusschock** [comment]
@vincentvaroquauxads thanks for checking it out and submitting this PR. Would you be interested in enabling the change with num_classes=17 and fixing the issues arising from this? It should almost always be dtype issues.

**Comment by vincentvaroquauxads** [comment]
I investigated a little bit more and some tests fail because `ignore_index=-1` and `target.dtype=uint8`.
I changed `target.dtype` to `torch.int8`, but I still have issues...

Actually, after hours, I finally realized that, using the original master branch, tests are broken in my environment.
It might be linked with this warning:
`sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.` (scikit-learn 1.0.2)

Sorry, no time to investigate further... And it's another issue (no link with this PR)...

I pushed the unit tests, if anyone can try them on a clean environment.

**Review by Borda — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]
